Adds support for ESP32S2 to the eps32_touch component.

larsenglund opened the pull request:
# What does this implement/fix? 

Adds support for ESP32S2 to the eps32_touch component

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other


## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
esphome:
  name: esphome-s2-test
  platformio_options:
      platform:
        - https://github.com/platformio/platform-espressif32.git#feature/arduino-upstream
      platform_packages:
        - framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git#2.0.2
        - platformio/tool-esptoolpy @ ~1.30100
        - espressif/toolchain-xtensa-esp32 @ 8.4.0+2021r2
      board_build.mcu: esp32s2

esp32:
  board: esp32dev
  variant: esp32s2

external_components:
  source: github://pr#3080 
  components:
    - esp32_touch
  refresh: 0s

# Enable logging
logger:

# Enable Home Assistant API
api:

ota:
  password: "ed54e3e66d1519b122a42c850e4faa8d"

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "esphome-s2-test"
    password: "vsv2QbrPo1Mz"

captive_portal:

esp32_touch:
  setup_mode: true

binary_sensor:
  - platform: esp32_touch
    name: "s2_touch_2"
    pin: GPIO2
    threshold: 1000

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Added ESP32S2 support

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Update binary_sensor.py

larsenglund committed:
Update binary_sensor.py

larsenglund committed:
Update binary_sensor.py

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Update binary_sensor.py

larsenglund committed:
Update binary_sensor.py

larsenglund committed:
Update esp32_touch.cpp

larsenglund committed:
Update esp32_touch.cpp

OttoWinter commented on esphome/components/esp32_touch/binary_sensor.py:
```diff
@@ -15,23 +20,42 @@
-    if value not in TOUCH_PADS:
-        raise cv.Invalid(f"Pin {value} does not support touch pads.")
+    variant = get_esp32_variant()
+    if value not in TOUCH_PADS[variant]:
```
We should check before that we support the given variant at all
```python
        if variant not in TOUCH_PADS:
            raise cv.Invalid(some_msg)
```

larsenglund replied:
Indeed, good suggestion.

larsenglund replied:
I added that.

OttoWinter commented on esphome/components/esp32_touch/esp32_touch.cpp:
```diff
@@ -14,17 +14,41 @@ void ESP32TouchComponent::setup() {
   touch_pad_init();
 
+#if defined(USE_ESP32) && !defined(USE_ESP32_VARIANT_ESP32S2)
   if (this->iir_filter_enabled_()) {
```
Is it correct that the esp32s2 does not support the IIR filter? If so, we should catch that in validation

larsenglund replied:
No, it does support both a common noise filter (which is enabled in my pr) and a IIR filter. The IIR for the ESP32S2 seems to be implemented differently than the ESP32 though. It takes in a struct with a bunch of parameters for the filter, the difference can be seen in the docs for the touchpad; 
ESP32S2: https://docs.espressif.com/projects/esp-idf/en/latest/esp32s2/api-reference/peripherals/touch_pad.html
ESP32: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/touch_pad.html
I got good results with just the common noise filter enabled so I didn't implement code to handle the IIR filter.

OttoWinter commented on esphome/components/esp32_touch/esp32_touch.cpp:
```diff
@@ -152,7 +186,11 @@ void ESP32TouchComponent::on_shutdown() {
 
       // No filter available when using as wake-up source.
+#if defined(USE_ESP32) && defined(USE_ESP32_VARIANT_ESP32S2)
+      touch_pad_config(child->get_touch_pad());
```
Same for wakeup threshold, is it supported on ESP32S2?

larsenglund replied:
It seems to support it by looking at the docs for the touchpad but it seems to be implemented slightly different and I had no need for it so I haven't implemented it.
ESP32S2: https://docs.espressif.com/projects/esp-idf/en/latest/esp32s2/api-reference/peripherals/touch_pad.html
ESP32: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/touch_pad.html

larsenglund committed:
Update binary_sensor.py

check before that we support the given variant at all

larsenglund committed:
Update esp32_touch.cpp

Touch sensor sends ON when not touched and OFF when touched for ESP32S2, this inverts that.

jesserockz commented on esphome/components/esp32_touch/binary_sensor.py:
```diff
@@ -15,23 +20,44 @@
+        raise cv.Invalid(f"{variant} not supported")
+    if value not in TOUCH_PADS[variant]:
+        raise cv.Invalid(f"{variant} doesn't support ADC on pin {value}")
     return value
```
I think if you change this, you don't need to look up the variant in to_code and can directly set my suggestion in my below comment. (Please do test this theory)
```suggestion
    return cv.enum(TOUCH_PADS[variant])(value)
```

larsenglund replied:
Thanks for the feedback! I got a bit sidetracked with other stuff but will fix this soon (if it's not already fixed in the main branch by now)

jesserockz commented on esphome/components/esp32_touch/binary_sensor.py:
```diff
@@ -52,9 +78,10 @@ def validate_touch_pad(value):
     var = cg.new_Pvariable(
         config[CONF_ID],
-        TOUCH_PADS[config[CONF_PIN]],
+        TOUCH_PADS[variant][config[CONF_PIN]],
```
As per above, you would get the `cg.global_ns.TOUCH_PAD_NUMx` value here directly like this.
```suggestion
        config[CONF_PIN],
```

larsenglund replied:
Thanks for the feedback! I got a bit sidetracked with other stuff but will fix this soon (if it's not already fixed in the main branch by now)

jesserockz reviewed (changes requested):
In the `esp32_touch.cpp` file, you don't need to have `#if defined(USE_ESP32)` as the first condition as the whole file is esp32 only.
